src/Build/Evaluation/SimpleProjectRootElementCache.cs

```diff
@@ -0,0 +1,127 @@
+        }
+
+        internal override ProjectRootElement Get(
+            string projectFile,
+            OpenProjectRootElement openProjectRootElement,
+            bool isExplicitlyLoaded,
+            bool? preserveFormatting)
+        {
+            // Should already have been canonicalized
+            ErrorUtilities.VerifyThrowInternalRooted(projectFile);
+
+            return openProjectRootElement == null
+                ? GetFromCache(projectFile)
+                : GetFromOrAddToCache(projectFile, openProjectRootElement);
+        }
+
+        private ProjectRootElement GetFromCache(string projectFile)
+        {
+            if (_cache.TryGetValue(projectFile, out ProjectRootElement projectRootElement))
+            {
+                return projectRootElement;
+            }
+
+            return null;
+        }
+
+        private ProjectRootElement GetFromOrAddToCache(string projectFile, OpenProjectRootElement openFunc)
+        {
+            return _cache.GetOrAdd(projectFile, key =>
+            {
+                var rootElement = openFunc(key, this);
+                ErrorUtilities.VerifyThrowInternalNull(rootElement, "projectRootElement");
+                ErrorUtilities.VerifyThrow(rootElement.FullPath.Equals(key, StringComparison.OrdinalIgnoreCase),
+                    "Got project back with incorrect path");
+                ErrorUtilities.VerifyThrow(_cache.TryGetValue(key, out _),
+                    "Open should have renamed into cache and boosted");
+
+                return rootElement;
+            });
+        }
+
+        internal override void AddEntry(ProjectRootElement projectRootElement)
+        {
+            if (_cache.TryAdd(projectRootElement.FullPath, projectRootElement))
+            {
+                RaiseProjectRootElementAddedToCacheEvent(projectRootElement);
+            }
+        }
+
+
+        internal override void RenameEntry(string oldFullPath, ProjectRootElement projectRootElement)
+        {
+            throw new NotImplementedException();
+        }
+
+        internal override ProjectRootElement TryGet(string projectFile)
+        {
+            return TryGet(projectFile, null);
+        }
+
+
+        internal override ProjectRootElement TryGet(string projectFile, bool? preserveFormatting)
+        {
+            return Get(
+                projectFile,
+                null,
+                false,
+                preserveFormatting);
+        }
+
+        internal override void DiscardStrongReferences()
+        {
+        }
+
+        internal override void Clear()
+        {
+            _cache.Clear();
+        }
+
+        internal override void DiscardImplicitReferences()
+        {
+            throw new NotImplementedException();
+        }
+
+        internal override void DiscardAnyWeakReference(ProjectRootElement projectRootElement)
+        {
+            ErrorUtilities.VerifyThrowArgumentNull(projectRootElement, "projectRootElement");
```

Comment:
Is this actually called during evaluations, or can it throw?

Reply:
Is there any value in `VerifyThrowArgumentNull` in a non-public API when it will only be hit if MSBuild itself has a bug and the very next line will give you a null reference exception instead?

Reply:
It's more about making the method contract explicit as executable documentation than anything else. Yes, things will eventually fail one way or another, but at least this way it's explicit.